prow/sidecar/run.go

```diff
@@ -38,6 +40,13 @@ import (
 // and then post the status of that process and any artifacts
 // to cloud storage.
 func (o Options) Run() error {
+	// If we are being asked to terminate by the kubelet we must
+	// hope that the test process exits cleanly and we can get a
+	// chance to upload artifacts to GCS. The only valid way for
+	// this program to exit after a SIGINT or SIGTERM is to finish
+	// uploading, so we ignore the signal.
+	signal.Ignore(os.Interrupt, syscall.SIGTERM)
```

Comment:
could we not opportunistically start uploading whatever artifacts we have upon receiving these signals?

Reply:
What should we do if we upload some partial result (or, for instance, scan a directory as empty) at `SIGTERM` but then we observe the test process exit? Cancel all pending uploads and rescan?

Reply:
bandwidth isn't usually a concern for these I think (?), so my first thought is that as long as we upload in parallel we can rescan and upload [on process exit] without cancelling the previous ones [from SIGTERM]. 
sure it's racy, but if we cancel the upload we might not finish posting anything, where as if the rescan finishes uploading I think the likely scenario is those uploads [after process exit] will override the SIGTERM uploads. 
and GCS will only replace an already uploaded entry ~ atomically if we actually mark the upload complete, so if the rescan starts to upload something but never finishes we hopefully still finished the one from SIGTERM.

I definitely think it's more useful to guarantee we upload whatever we can when pods are less-than-gracefully terminated than nothing.

If we observe SIGTERM after we've already observed process exit somehow, we should not do anything of course.